focus-android/app/src/main/java/org/mozilla/focus/cookiebannerreducer/CookieBannerReducerMiddleware.kt

```diff
@@ -0,0 +1,203 @@
+
+    override fun invoke(
+        context: MiddlewareContext<CookieBannerReducerState, CookieBannerReducerAction>,
+        next: (CookieBannerReducerAction) -> Unit,
+        action: CookieBannerReducerAction,
+    ) {
+        when (action) {
+            is CookieBannerReducerAction.InitCookieBannerReducer -> {
+                /**
+                 * The initial CookieBannerReducerState when the user enters first in the screen
+                 */
+                initCookieBannerReducer(context)
+            }
+
+            is CookieBannerReducerAction.ToggleCookieBannerException -> {
+                handleCookieBannerToggle(action, context)
+                next(action)
+            }
+            is CookieBannerReducerAction.RequestReportSite -> {
+                reportSite(action, context)
+                next(action)
+            }
+            else -> {
+                next(action)
+            }
+        }
+    }
+
+    private fun handleCookieBannerToggle(
+        action: CookieBannerReducerAction.ToggleCookieBannerException,
+        context: MiddlewareContext<CookieBannerReducerState, CookieBannerReducerAction>,
+    ) {
+        ioScope.launch {
+            if (action.isCookieBannerHandlingExceptionEnabled) {
+                cookieBannersStorage.removeException(currentTab.content.url, true)
+                CookieBanner.exceptionRemoved.record(NoExtras())
+                context.store.dispatch(
+                    CookieBannerReducerAction.UpdateCookieBannerReducerStatus(
+                        CookieBannerReducerStatus.NoException,
+                    ),
+                )
+            } else {
+                clearSiteData()
+                cookieBannersStorage.addPersistentExceptionInPrivateMode(currentTab.content.url)
+                CookieBanner.exceptionAdded.record(NoExtras())
+                context.store.dispatch(
+                    CookieBannerReducerAction.UpdateCookieBannerReducerStatus(
+                        CookieBannerReducerStatus.HasException,
+                    ),
+                )
+            }
+            appContext.components.sessionUseCases.reload()
+        }
+    }
+
+    private fun reportSite(
+        action: CookieBannerReducerAction.RequestReportSite,
+        context: MiddlewareContext<CookieBannerReducerState, CookieBannerReducerAction>,
+    ) {
+        CookieBanner.reportSiteDomain.set(action.siteToReport)
+        CoroutineScope(Dispatchers.IO).launch {
+            Pings.cookieBannerReportSite.submit()
+            context.store.dispatch(
+                CookieBannerReducerAction.ShowSnackBarForSiteToReport(
+                    true,
+                ),
+            )
+            context.store.dispatch(
+                CookieBannerReducerAction.UpdateCookieBannerReducerStatus(
+                    CookieBannerReducerStatus.CookieBannerUnsupportedSiteRequestWasSubmitted,
+                ),
+            )
```

Comment:
I'm not familiar with pings, but I guess we would like to keep track of the reported sites, should we include the site as part of the ping?

Reply:
This is an answer from @travis79 . He helped me with this issue. 
Yes, Glean can record all of those things, in fact, every Glean ["ping"](https://mozilla.github.io/glean/book/user/pings/index.html) already contains much of this information.
Timestamp:  Glean has a few different types of these, including timestamps built into every ping to indicate the beginning and end of the window of time from which the data in the ping was collected. If you need a specific timestamp for something, Based on the doc you linked, I think that the built-in Glean timestamps combined with the use of a "custom" ping should probably get you what you want.
Domain: This is definitely Category 3/4 (highly-sensitive) data, and Glean provides a [URL](https://mozilla.github.io/glean/book/reference/metrics/url.html) data type for this sort of data. On top of that, sending this in a custom ping will help to isolate it from other telemetry and allow for a custom data-retention policy for it. Finally, using a custom ping will allow you to not include the telemetry client_id, which can further help to protect the user's privacy by making it harder to link to other data we may collect.
Country: We collect this information when the ping is sent, automatically from a geo-lookup using the IP address before we discard the IP address to protect privacy. Unless you are prompting the user for this information and cannot rely on the automated geo information. In that case, using a String data type to collect this if it is user input would be advisable.
Language/Locale: Just like country, Glean collects this information from the device automatically based on the user's settings. If you need to record something other than the automatic locale that Glean collects, a String data type would also be advisable to use for this.
Here is an example of a custom ping defined in Fenix that does not send the telemetry client-id with the custom ping data: https://github.com/mozilla-mobile/firefox-android/blob/8cf9723866b8343423cc5ac6112319dab7cc6be2/fenix/app/pings.yaml#L34
And here is an example from Fenix's metrics.yaml file of a metric sent in that "topsites-impression" custom ping. It's a URL that would be similar to the domain you wish to capture: https://github.com/mozilla-mobile/firefox-android/blob/main/fenix/app/metrics.yaml#L5942

Reply:
Thanks for the context, I think we could add @travis79 as reviewer to verify this implementation and for data review as this is Category 3/4 (highly-sensitive) data.

Reply:
The implementation looks good, separating the url data from any identifiers by use of a custom ping is preferred when sending something that could be sensitive like a URL, even if that url was submitted explicitly by the user.